### PR TITLE
Auto publish parser js to npm

### PR DESCRIPTION
**Pull request by tanishiking**
This PR

- Introduce https://github.com/davenverse/sbt-npm-package
- Setup GitHub action workflow to automate release parsersJS
  - When we tag like `v4.1.6`, the workflow will publish parsersJS under version 4.1.6
  - e.g. https://github.com/tanishiking/scalameta/actions/runs/3534906546/jobs/5932323432
  - published artifact

**TODO**

- We have to set repository secret `NPM_TOKEN` with npm access token generated by someone who is a member of `scalameta-parsers` (@tanishiking @gabro has access AFAIK), and the type should be "automated".
  - Can someone give me the right to modify the repository secret? or do that instead of me?

---

https://github.com/davenverse/sbt-npm-package provides `sbt parsersJS/npmPackagePublishg` task and it does

- `sbt parsersJS/npmPackageNpmrc` to write `scalameta/parsers/js/target/scala-2.13/npm-package/.npmrc` with content `//registry.npmjs.org/:_authToken=${NPM_TOKEN}`
- Write `scalameta/parsers/js/target/scala-2.13/npm-package/package.json` based on sbt settings
- Copy `parser-opt.js` to `.../npm-package/main.js` (+ `main.js.map` too)
- Copy (specified) README.md to the `.../npm-package/README.md`
- Run `npm publish` on the `npm-package` directory

Here's a test published parser-js library https://www.npmjs.com/package/@tanishiking/scalameta-parsers/v/4.6.1

Generated `package.json` looks like

```json
{
  "name" : "@tanishiking/scalameta-parsers",
  "description" : "Library to parse Scala programs",
  "version" : "4.6.0",
  "type" : "commonjs",
  "repository" : {
    "type" : "git",
    "url" : "https://github.com/scalameta/scalameta"
  },
  "author" : "scalameta",
  "license" : "BSD-3-Clause",
  "main" : "main.js",
  "dependencies" : {
    
  },
  "devDependencies" : {
    
  },
  "keywords" : [
    "scala",
    "parser"
  ],
  "homepage" : "https://scalameta.org/"
}
```

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -216,9 +216,24 @@ lazy val parsers = crossProject(JSPlatform, JVMPlatform, NativePlatform)
...
     })
   )
-  .jsSettings(
+  .jsConfigure(
+      _.enablePlugins(NpmPackagePlugin)
+  ).jsSettings(
     commonJsSettings,
-    scalaJSLinkerConfig ~= { _.withModuleKind(ModuleKind.CommonJSModule) }
+    scalaJSLinkerConfig ~= { _.withModuleKind(ModuleKind.CommonJSModule) },
+    npmPackageVersion := "4.6.1",
```
Comment (tanishiking) [inline build.sbt:224]: We'll remove this field by default, `version` in sbt will be used :+1: (This is why I'm using `sbt-npm-package`)

https://github.com/davenverse/sbt-npm-package/blob/a638ce6933f39a02ebe721124229e27dba6d63a0/core/src/main/scala/io/chrisdavenport/npmpackage/sbtplugin/NpmPackagePlugin.scala#L204-L209

File: build.sbt [hunk 2]
```diff
@@ -216,9 +216,24 @@ lazy val parsers = crossProject(JSPlatform, JVMPlatform, NativePlatform)
...
+    npmPackageName := "@tanishiking/scalameta-parsers",
```
Comment (tanishiking) [inline build.sbt:225]: Change to `scalameta-parsers` later

File: build.sbt [hunk 3]
```diff
@@ -216,9 +216,24 @@ lazy val parsers = crossProject(JSPlatform, JVMPlatform, NativePlatform)
...
+    npmPackageRepository := Some("https://github.com/scalameta/scalameta"),
+    npmPackageAuthor := "scalameta",
+    npmPackageLicense := Some("BSD-3-Clause"),
+    npmPackageKeywords := Seq("scala", "parser"),
+    npmPackageStage := org.scalajs.sbtplugin.Stage.FullOpt,
+    npmPackageAdditionalNpmConfig := Map(
+       "homepage" -> _root_.io.circe.Json.fromString("https://scalameta.org/"),
+    ),
+    npmPackageREADME := Some(file("README.npm.md")),
+    npmPackageNpmExtraArgs := Seq("--access=public")
```
Comment (tanishiking) [inline build.sbt:236]: Don't need this once we rename from `@tanishiking/scalameta-parsers` to `scalameta-parsers`

File: .github/workflows/release-js.yml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+      - uses: actions/setup-node@v2
+        with:
+          node-version: "18"
+      - uses: actions/checkout@v3
+      - uses: olafurpg/setup-scala@v13
+      - run: git fetch --unshallow
+      - name: Publish
+        run: sbt parsersJS/npmPackagePublish
+        env:
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}"
```
Comment (tanishiking) [inline .github/workflows/release-js.yml:19]: Not yet set the secret

File: .github/workflows/release-js.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+name: Release
+on:
+  push:
+    branches: [main]
+    tags: ["*"]
+jobs:
+  publish:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-node@v2
```
Comment (xuwei-k) [inline .github/workflows/release-js.yml:10]:
```suggestion
      - uses: actions/setup-node@v3
```

**Comment by tanishiking** [comment]
Sorry for taking so long to finish up the PR, but now it's ready for review :)

**Review by gabro — APPROVED** [review]
Great job! 🥳

**Review by tgodzik — APPROVED** [review]
LGTM! Let's try it out with the new release!

**Comment by tanishiking** [comment]
Thanks! but we have to set repository secret, can someone add it or give me an access right to add secret?

**Comment by tanishiking** [comment]
Added `NPM_TOKEN` repository secret :+1:

**Pull request merged**